lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,71 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp component, OpBuilder builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp invokeOp);
+  ComponentOp component;
+  OpBuilder builder;
+  // It is used to pass to the go port and assign a value to the go port.
+  hw::ConstantOp constantOp = nullptr;
+  llvm::StringMap<size_t> groupNameMap;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  llvm::SmallVector<InvokeOp> invokeOps =
+      component.getControlOp().getInvokeOps();
+  for (InvokeOp op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp invokeOp) {
+  // Create a ConstantOp to assign a value to the go port.
+  if (!constantOp) {
+    Operation *prevNode = component.getWiresOp().getOperation()->getPrevNode();
+    builder.setInsertionPointAfter(prevNode);
+    constantOp = builder.create<hw::ConstantOp>(prevNode->getLoc(),
+                                                builder.getI1Type(), 1);
+  }
+  Location loc = component.getWiresOp().getLoc();
+  // Set the insertion point at the end of the wires block.
+  builder.setInsertionPointToEnd(component.getWiresOp().getBodyBlock());
+  llvm::StringRef callee = invokeOp.getCallee();
+  if (groupNameMap.find(callee) == groupNameMap.end())
+    groupNameMap[callee] = 0;
+  std::string groupName =
+      "invoke_" + callee.str() + "_" + std::to_string(groupNameMap[callee]);
+  // Check if the group name is correct.
+  while (component.getWiresOp().lookupSymbol(groupName))
+    groupName =
+        "invoke_" + callee.str() + "_" + std::to_string(++groupNameMap[callee]);
```

Comment:
This seems pretty inefficient; can we avoid generating a new string each time here to guarantee this is a unique name? Aren't we guaranteed this already by the index in `groupNameMap[callee]` ?

Reply:
Initially I had this in mind, in fact initially I had the idea of a group with the name `invoke_lable`,`lable` is a number, but then I remembered that it could lead to name conflicts, so I added the code to look it up in the symbol table, but then I thought that if I just used lable, it would lead to less readability, so I have the code I have now.
* Aren't we guaranteed this already by the index in groupNameMap[callee] ? 
Assuming the user has written `calyx.group invoke_id_1` in `calyx.wires`,but using `invoke @id`,then the name of the first group generated in the pass should be `invoke_id_0`,but the name of the subsequent groups can no longer be `invoke_id_1`,I have considered such a situation. (Maybe I'm overthinking it?)

Reply:
I did a simple optimization here .